Firestore/Source/Util/FSTDispatchQueue.mm

```diff
@@ -21,8 +21,130 @@
++ (instancetype)createAndScheduleWithQueue:(FSTDispatchQueue *)queue
+                                   timerID:(FSTTimerID)timerID
+                                     delay:(NSTimeInterval)delay
+                                  callback:(void (^)(void))callback;
+
+/**
+ * Queues the operation to run immediately (if it hasn't already been run or canceled).
+ */
+- (void)skipDelay;
+
+@end
+
+@implementation FSTDelayedCallback
+
+- (instancetype)initWithQueue:(FSTDispatchQueue *)queue
+                      timerID:(FSTTimerID)timerID
+                   targetTime:(NSTimeInterval)targetTime
+                     callback:(void (^)(void))callback {
+  if (self = [super init]) {
+    _queue = queue;
+    _timerID = timerID;
+    _targetTime = targetTime;
+    _callback = callback;
+    _pending = YES;
+  }
+  return self;
+}
+
++ (instancetype)createAndScheduleWithQueue:(FSTDispatchQueue *)queue
+                                   timerID:(FSTTimerID)timerID
+                                     delay:(NSTimeInterval)delay
+                                  callback:(void (^)(void))callback {
+  NSTimeInterval targetTime = [[NSDate date] timeIntervalSince1970] + delay;
+  FSTDelayedCallback *delayedCallback = [[FSTDelayedCallback alloc] initWithQueue:queue
+                                                                          timerID:timerID
+                                                                       targetTime:targetTime
+                                                                         callback:callback];
+  [delayedCallback startWithDelay:delay];
+  return delayedCallback;
+}
+
+/**
+ * Starts the timer. This is called immediately after construction by createAndScheduleWithQueue.
+ */
+- (void)startWithDelay:(NSTimeInterval)delay {
+  dispatch_time_t delayNs = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delay * NSEC_PER_SEC));
+  dispatch_after(delayNs, self.queue.queue, ^{
+    [self delayDidElapse];
+  });
+}
+
+- (void)skipDelay {
+  [self.queue dispatchAsyncAllowingSameQueue:^{
+    [self delayDidElapse];
+  }];
+}
+
+- (void)cancel {
+  [self.queue verifyIsCurrentQueue];
+  if (self.isPending) {
+    // PORTING NOTE: There's no way to actually cancel the dispatched callback, but it'll be a no-op
+    // since we set pending to NO.
+    [self clear];
+  }
+}
+
+- (void)delayDidElapse {
+  [self.queue verifyIsCurrentQueue];
+  if (self.isPending) {
+    [self clear];
+    self.callback();
+  }
+}
+
+/**
+ * Marks this delayed callback as no longer pending, and notifies the FSTDispatchQueue that it
+ * should be removed.
+ */
+- (void)clear {
+  self.pending = NO;
+  [self.queue removeDelayedCallback:self];
+}
+
+@end
+
+#pragma mark - FSTDispatchQueue
+
 @interface FSTDispatchQueue ()
+
+/**
+ * Callbacks scheduled to be queued in the future. Callbacks are automatically removed after they
+ * are run or canceled.
+ */
+@property(nonatomic, strong, readonly) NSMutableArray<FSTDelayedCallback *> *delayedCallbacks;
```

Comment:
You could simplify some of the code in this class if this was a Dictionary from TimeId to FSTDelayedCallback.

Reply:
Interesting point. I think I'd rather not since: 1) I didn't do that on Web, 2) If we ever need to allow multiple instances of a TimerID we'd need to go back to an array, and 3) it's convenient to just sort this and enumerate it in the runDelayedCallbacks method...